src/libraries/Microsoft.Extensions.Options/src/OptionsCache.cs

```diff
@@ -31,14 +31,47 @@ public class OptionsCache<[DynamicallyAccessedMembers(Options.DynamicallyAccesse
         public virtual TOptions GetOrAdd(string? name, Func<TOptions> createOptions!!)
         {
             name ??= Options.DefaultName;
-            Lazy<TOptions>? value;
+            Lazy<TOptions> value;
 
-#if NETSTANDARD2_1
+#if NET || NETSTANDARD2_1
             value = _cache.GetOrAdd(name, static (name, createOptions) => new Lazy<TOptions>(createOptions), createOptions);
 #else
             if (!_cache.TryGetValue(name, out value))
             {
-                value = _cache.GetOrAdd(name, new Lazy<TOptions>(createOptions));
+                // copying captured variables to locals avoids allocating a closure if we don't enter the if
+                Func<TOptions> localCreateOptions = createOptions;
+                value = _cache.GetOrAdd(name, new Lazy<TOptions>(localCreateOptions));
+            }
+#endif
+
+            return value.Value;
+        }
+
+        internal TOptions GetOrAdd<TArg>(string? name, Func<string, TArg, TOptions> createOptions, TArg factoryArgument)
+        {
+            // For compatibility, fall back to public GetOrAdd() if we're in a derived class.
+            if (GetType() != typeof(OptionsCache<TOptions>))
+            {
+                // copying captured variables to locals avoids allocating a closure if we don't enter the if
+                string? localName = name;
+                Func<string, TArg, TOptions> localCreateOptions = createOptions;
+                TArg localFactoryArgument = factoryArgument;
+                return GetOrAdd(name, () => localCreateOptions(localName ?? Options.DefaultName, localFactoryArgument));
+            }
+
+            name ??= Options.DefaultName;
+            Lazy<TOptions> value;
+
+#if NET || NETSTANDARD2_1
+            value = _cache.GetOrAdd(name, static (name, arg) => new Lazy<TOptions>(arg.createOptions(name, arg.factoryArgument)), (createOptions, factoryArgument));
+#else
+            if (!_cache.TryGetValue(name, out value))
```

Comment:
On .NET Framework and netstandard2.0, why don't we just fall back to the public GetOrAdd() as well? I don't think we need to try making a fast-path outside of .NETCoreApp. Especially considering how complicated this method is getting.

Reply:
Makes sense. Wasn't sure how much we should care about this sort of thing on the older frameworks.